### PR TITLE
Schema registry resource

### PR DESCRIPTION
**Pull request by dmicic**

### CODE REVIEW

**Comment by dmicic** [comment]
@Mongey My Ubuntu subsystem dies while building and can't test it properly on my Win machine, although compilation works.
Can you give it a try maybe?

**Comment by rjudin** [comment]
seems Github Action stuck for the last 3 days on the required steps for verification.

image


https://github.com/Mongey/terraform-provider-confluentcloud/blob/e690efc76ad15550fdbd0279d4ab9895c611f4ec/.github/workflows/snapshot.yml#L9

https://github.com/Mongey/terraform-provider-confluentcloud/blob/e690efc76ad15550fdbd0279d4ab9895c611f4ec/.github/workflows/test.yml#L9

---

@Mongey could you please review this PR and merge?

Thank you!

**Review by rjudin — APPROVED** [review]
👌  looks good to me.

**Comment by dmicic** [comment]
> @Mongey could you please review this PR and merge?

We need the [go client PR](https://github.com/cgroschupp/go-client-confluent-cloud/pull/11) to be merged first.

**Comment by Mongey** [comment]
> seems Github Action stuck for the last 3 days on the required steps for verification.


🤞 this works on next push, given #44

File: ccloud/provider.go [hunk 1]
```diff
@@ -30,6 +30,7 @@ func Provider() terraform.ResourceProvider {
 			"confluentcloud_api_key":         apiKeyResource(),
 			"confluentcloud_environment":     environmentResource(),
 			"confluentcloud_service_account": serviceAccountResource(),
+			"confluentcloud_schema_registry": schemaRegistryResource(),
```
Comment (Mongey) [inline ccloud/provider.go:33]: 🔤 
```suggestion
			"confluentcloud_schema_registry": schemaRegistryResource(),
			"confluentcloud_service_account": serviceAccountResource(),
```

File: ccloud/resource_schema_registry.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+	err = d.Set("environment_id", environment)
+	if err != nil {
+		err = d.Set("endpoint", env.Endpoint)
+	}
+
+	return err
+}
+
+func schemaRegistryDelete(d *schema.ResourceData, meta interface{}) error {
+	log.Printf("[INFO] Schema registry cannot be deleted: %s", d.Id())
```
Comment (Mongey) [inline ccloud/resource_schema_registry.go:88]: 🤔 How come?

Comment (dmicic) [reply ccloud/resource_schema_registry.go:88]: The official ccloud API supports only the “enable” function, no delete. That is also consitent with the UI. Even if there is a way to disable the registry again, I would not how as there is no reference anywhere.

**Comment by dmicic** [comment]
The test does not fail properly. Following error has been raised. Suggest to merge this on here now.. https://github.com/cgroschupp/go-client-confluent-cloud/pull/11

```
Error: ccloud/resource_schema_registry.go:54:15: c.CreateSchemaRegistry undefined (type *confluentcloud.Client has no field or method CreateSchemaRegistry)
Error: ccloud/resource_schema_registry.go:74:15: c.GetSchemaRegistry undefined (type *confluentcloud.Client has no field or method GetSchemaRegistry)
make: *** [test] Error 2
Error: GNUmakefile:55: recipe for target 'test' failed
```

**Comment by cgroschupp** [comment]
If this change is irreversible, is it not better to add a parameter to the confluentcloud_kafka_cluster resource?

**Comment by dmicic** [comment]
> If this change is irreversible, is it not better to add a parameter to the `confluentcloud_kafka_cluster` resource, like `enable_schema_registry`?

Thought about that as well. But the schema registry is not bound to a specific cluster. It's one registry per environment. And adding this extra parameter to the `confluentcloud_environment` won't work either because ccloud expects you to have at least one cluster to be deployed before you can enable the registry.

**Comment by dmicic** [comment]
Question from a Golang beginner. Do I need to update/refresh the go.sum file in order to fetch the latest version of the go confluent client lib?

**Comment by Mongey** [comment]
@dmicic yep you've got to run `go get -u github.com/cgroschupp/go-client-confluent-cloud` to update the lib

**Comment by dmicic** [comment]
@Mongey thanks! worked. Now only the private key issue remaining.

**Comment by dmicic** [comment]
@Mongey Can you fix the GPG key please?

**Comment by Mongey** [comment]
@dmicic can you rebase ?

**Comment by dmicic** [comment]
@Mongey That should do the job now. 🎉

**Comment by rjudin** [comment]
@dmicic, @Mongey thank you!
What is left to specify Kafka topic schema?

**Comment by dmicic** [comment]
@rjudin you mean management of schemas (CRUD)? Not sure this should be part of a terraform module. Depends how somebody structures his deployment pipelines. 

@Mongey can you release the new changes, please? thanks

**Comment by rjudin** [comment]
> @rjudin you mean management of schemas (CRUD)? Not sure this should be part of a terraform module. Depends how somebody structures his deployment pipelines.

I meant resource `kafka_schema` for [Kafka provider](https://github.com/Mongey/terraform-provider-kafka)
like [this example](https://github.com/aiven/terraform-provider-aiven/blob/5853529826b49d30699f3da338f9add90b26fe79/examples/kafka_schemas/kafka_schema.tf#L9-L29)

**Comment by Mongey** [comment]
@rjudin There's probably another provider needed for interacting with the schema registry. https://github.com/search?q=terraform-provider+schema-registry

**Comment by dmicic** [comment]
I am not quite convinced that schema should be managed in the (infrastructure) code. Typically, you would want this to be closer to your solution, which might have different deployment lifecycle/frequency. 
 But that's only my preference maybe :)

**Comment by rjudin** [comment]
agree it is about app (not infra) 
We did (pipeline) separation by provider already:
- infra: confluent cloud provision (env, SAs and API keys, clusters) 
- app: Kafka topics and acls
---
Schema naturally needs to be managed, but required resource definition and separate keys. Did anynone do this already for confluent cloud schema registry?

**Comment by dmicic** [comment]
I see. Provider separation makes sense. 
I did this only using the ccloud CLI as part of my pipelines (eg uploading the schema files). But without change detection etc..

**Comment by dmicic** [comment]
@Mongey Any chance to deploy this feature any time soon? THx

**Pull request merged**